### PR TITLE
fix issues on nodes without lvm

### LINKED ISSUE #307: v2.0.3 is broken
## Describe the Bug
Updating to v2.0.3 causes failures:

```
2023-07-11 13:55:15 +0000 Facter (err): error while resolving custom facts in /opt/puppetlabs/puppet/cache/lib/facter/lvm_support.rb undefined method `message' for #<String:0x00000000047ec120>
```

Agent is RHEL8, lvm is not installed

### PR DESCRIPTION
**Pull request by rtib**
tested manually on nodes with and without lvm

Closes #307

### CODE REVIEW

**Comment by vchepkov** [comment]
According to https://www.puppet.com/docs/puppet/7/fact_overview.html, there should not be code outside setcode

```
Note: Set all code inside the sections outlined above ⁠— there must not be any code outside setcodeand confine blocks other than an optional has_weight statement in a custom fact.
```

**Comment by rtib** [comment]
That's true and one more cause why these legacy facts should be deprecated, as pointed out in Slack discussion by @cyberious
> Just note that when puppetlabs would be strict the lvm_vg_*_pvs, lvm_vg_* and lvm_pv_* facts should be marked as legacy facts and which are per default disabled with puppet 8. So maybe inventing time to use the new structured facts would be better in the long run.

They just need to work as long there is no alternative source for the particular information they provide and some time to migrate.

File: lib/facter/lvm_support.rb [hunk 1]
```diff
@@ -17,47 +17,48 @@
...
   confine lvm_support: true
 
-  vgs = Facter::Core::Execution.execute('vgs -o name --noheadings 2>/dev/null', timeout: 30)
-
-  if vgs.nil?
-    setcode { 0 }
-  else
-    vg_list = vgs.split
-    setcode { vg_list.length }
+  if Facter.value(:lvm_support)
```
Comment (bastelfreak) [inline lib/facter/lvm_support.rb:20]: this if should not be required, the `lvm_vgs` fact is already confineed on `lvm_support` (see line 18).

File: lib/facter/lvm_support.rb [hunk 2]
```diff
@@ -17,47 +17,48 @@
...
 Facter.add('lvm_pvs') do
   confine lvm_support: true
 
-  pvs = Facter::Core::Execution.execute('pvs -o name --noheadings 2>/dev/null', timeout: 30)
-  if pvs.nil?
-    setcode { 0 }
-  else
-    pv_list = pvs.split
-    setcode { pv_list.length }
+  if Facter.value(:lvm_support)
```
Comment (bastelfreak) [inline lib/facter/lvm_support.rb:50]: same here

File: lib/facter/lvm_support.rb (hunk 1 above)
Comment (rtib) [reply lib/facter/lvm_support.rb:20]: indeed, if it would be within the setcode block; I'm currently busy with getting rid off these nasty pieces here and to find a solution which works and passes spec tests

**Comment by rtib** [comment]
The thing is as follows:

TL;DR
Those facts with programmatically generated names, i.e. `lvm_vg_<n>` and `lvm_vg_<vg_name>_pvs` rely on code outside any setcode blocks, but require the results of commands executed before. It is not possible to set global variables within a setcode block, that's why those commands need to be executed outside, where they are not guarded by confine statements. This obviously violates the requirement to have all code of this kind within setcode or confine blocks. They could, however, be accessed as fact values, but as they are needed to produce the names of the above facts, that would violate the requirement to have such code within setcode blocks only. Current RSpec implementation isn't able to deal with code outside setcode and confine statements, thus these features cannot be tested within spec. 

To me, all that sums up as the code as it is works for now, but cannot be tested thoroughly, and might break in near future. Any code, that can be tested and is future prove, won't be able to produce fact names programmatically - dispute me if I'm wrong.

I'd propose the following solution:
* leave the code within `lvm_support.rb` as it is in this PR, manually tested
* drop the failing spec tests
* deprecate the legacy facts having programmatically generated names now, and remove them as soon as possible
* implement replacement using a modern approach, i.e. structured/aggregated facts reporting the relations between pvs, vgs and lvs

File: lib/facter/lvm_support.rb (hunk 2 above)
Comment (rtib) [reply lib/facter/lvm_support.rb:50]: Please have a look at https://github.com/puppetlabs/puppetlabs-lvm/pull/308#issuecomment-1633927597

**Comment by rtib** [comment]
That would work in the wild, though, it doesn't solve any problems we are facing here:

1. violating the format as described here: https://www.puppet.com/docs/puppet/7/fact_overview.html#writing_facts_simple_resolutions-how-to-format-facts
2. cannot write proper spec tests for

Moreover, `lvm_pv_<n>` can be replaced by using `physical_volumes`. The main focus is on the contents of `lvm_vg_<vg_name>_pvs`, which doesn't have an alternative source yet.

File: lib/facter/lvm_support.rb [hunk 3]
```diff
@@ -17,47 +17,48 @@
...
+    vgs = Facter::Core::Execution.execute('vgs -o name --noheadings 2>/dev/null', timeout: 30)
+    vg_list = vgs.split unless vgs.nil?
   end
+
+  setcode { vg_list.length }
 end
 
-# lvm_vg_[0-9]+
-#   VG name by index
+# # lvm_vg_[0-9]+
```
Comment (malikparvez) [inline lib/facter/lvm_support.rb:28]: any specific reason for double hashes?

Comment (rtib) [reply lib/facter/lvm_support.rb:28]: No. It's the result of an IDE accident.

**Comment by elfranne** [comment]
Tested on Windows 2019 too and works fine.

**Comment by vchepkov** [comment]
I defer to experts in Puppet's internals. As I stated earlier, v2.0.3 and updated code doesn't conform to the documentation, code in v2.0.2 did.

File: README.md [hunk 4]
```diff
@@ -346,6 +346,13 @@ Logical volume size can be extended, but not reduced -- this is for
 safety, as manual intervention is probably required for data
 migration, etc.
 
+## Deprecation Notice
+
+Some facts reported by this module are being deprecated in favor of upcomming structured facts.  The following facts are being deprecated:
```
Comment (Ramesh7) [inline README.md:351]: The below facts are the custom flat facts, so not sure if these are gonna deprecate as well?
AFAIK the custom facts (irrespective of flat or structured facts) these will not be deprecated. Only the [legacy facts](https://www.puppet.com/docs/puppet/7/core_facts.html#legacy-facts) which will get deprecated.

Comment (rtib) [reply README.md:351]: Facter code generating fact names dynamically on the fly is problematic: violates coding rules, hard to maintain, even harder to test. That's why we should get rid of them.

**Comment by crazymind1337** [comment]
Can we please have this released soon?

**Review by Ramesh7 — APPROVED** [review]
we are moving with this change for now but for long term we need to review these facts with manifest use-case and update accordingly.

**Comment by rajat-puppet** [comment]
These changes have been released as part of `2.1.0` release version

**Comment by rtib** [comment]
Thank You.
